ampd/src/mvx/verifier.rs

```diff
@@ -0,0 +1,630 @@
+        let sender = unwrap_err_or_false!(STANDARD.decode(sender));
+        if sender.len() != 32 || &sender[0..32] != &msg.source_address.to_bytes() {
+            return false;
+        }
+
+        let destination_chain = unwrap_or_false!(topics.get(2));
+        let destination_chain = unwrap_err_or_false!(STANDARD.decode(destination_chain));
+        let destination_chain = unwrap_err_or_false!(String::from_utf8(destination_chain));
+        if destination_chain != msg.destination_chain.to_string() {
+            return false;
+        }
+
+        let destination_address = unwrap_or_false!(topics.get(3));
+        let destination_address = unwrap_err_or_false!(STANDARD.decode(destination_address));
+        let destination_address = unwrap_err_or_false!(String::from_utf8(destination_address));
+        if destination_address != msg.destination_address {
+            return false;
+        }
+
+        let payload_hash = unwrap_or_false!(topics.get(4));
+        let payload_hash = unwrap_err_or_false!(STANDARD.decode(payload_hash));
+        if payload_hash.len() != 32 || Hash::from_slice(payload_hash.as_slice()) != msg.payload_hash
+        {
+            return false;
+        }
+
+        return true;
+    }
+}
+
+impl PartialEq<VerifierSetConfirmation> for &Events {
+    fn eq(&self, verifier_set: &VerifierSetConfirmation) -> bool {
+        if self.identifier != ROTATE_SIGNERS_IDENTIFIER {
+            return false;
+        }
+
+        let topics = unwrap_or_false!(self.topics.as_ref());
+
+        let event_name = unwrap_or_false!(topics.get(0));
+        let event_name = unwrap_err_or_false!(STANDARD.decode(event_name));
+        if event_name.as_slice() != SIGNERS_ROTATED_EVENT.as_bytes() {
+            return false;
+        }
+
+        let signers_hash = unwrap_or_false!(topics.get(2));
+        let signers_hash = unwrap_err_or_false!(STANDARD.decode(signers_hash));
+
+        let data = unwrap_or_false!(self.data.as_ref());
+        let data = unwrap_err_or_false!(STANDARD.decode(data));
+
+        let weighted_signers = WeightedSigners::from(&verifier_set.verifier_set);
+
+        if signers_hash.len() != 32 || signers_hash.as_slice() != weighted_signers.hash().as_slice()
+        {
+            return false;
+        }
+
+        let encoded = unwrap_err_or_false!(weighted_signers.encode());
+        if encoded != data {
+            return false;
+        }
+
+        return true;
+    }
+}
+
+fn find_event<'a>(
+    transaction: &'a TransactionOnNetwork,
+    gateway_address: &Address,
+    log_index: usize,
+    identifier: &str,
+    needed_event_name: &[u8],
+) -> Option<&'a Events> {
+    if transaction.logs.is_none() {
+        return None;
+    }
+
+    // Because of current relayer limitation, if log_index is 0, we will search through the logs and get the first log
+    // which corresponds to the gateway address, hence only supporting one cross chain call per transaction

```

Comment:
Doesn't this mean I can verify the same message twice? Meaning I pass log index of 0, and then I pass it again with the actual log index.

Reply:
Hm, you are right, this is a possibility and may not be intended.

Then for now we will only support log index 0, and in the future when multiple such events will be supported on relayer side correctly we can update the ampd code to remove this search.

Although I think then we will also need to have some sort of logic to not be able to handle older transactions before the log index is supported.

Anyways, I don't expect this issue to be solved on the relayer/blockchain side, so we can live with the 1 callContract per transaction limitation.

Reply:
Fixed